File: src/services/danmaku.ts
import WebSocket from 'ws';
import { EventEmitter } from 'events';

interface DanmakuMessage {
  nickname: string;
  content: string;
  level: string;
  badgeName: string;
  badgeLevel: string;
}

class DanmakuClient extends EventEmitter {
  private roomId: string;
  private ws: WebSocket | null = null;
  private heartbeatInterval: NodeJS.Timeout | null = null;
  private reconnectAttempts: number = 0;
  private readonly maxReconnectAttempts: number = 5;
  private readonly heartbeatInterval_ms: number = 45000;
  private readonly wsBaseUrl = 'wss://danmuproxy.douyu.com:';
  private readonly port = 8503;
  private isDestroyed: boolean = false;
  private isConnecting: boolean = false;
  private readonly connectionTimeout_ms: number = 4000;

  constructor(roomId: string) {
    super();
    this.roomId = roomId;
  }

  private encodeMsg(msg: string): Buffer {
    const msgBytes = Buffer.from(msg);
    const packetLen = msgBytes.length + 9;
    
    const header = Buffer.alloc(12);
    header.writeInt32LE(packetLen, 0);
    header.writeInt32LE(packetLen, 4);
    header.writeInt16LE(689, 8);
    header.writeInt8(0, 10);
    header.writeInt8(0, 11);

    return Buffer.concat([header, msgBytes, Buffer.from([0])]);
  }

  private decodeMsg(data: Buffer): Record<string, string> | null {
    try {
      const content = data.slice(12, -1).toString('utf-8');
      const result: Record<string, string> = {};
      
      content.split('/').forEach(item => {
        if (!item) return;
        const [key, value] = item.split('@=');
        if (key && value) {
          result[key] = value.replace(/@S/g, '/').replace(/@A/g, '@');
        }
      });
      
      return result;
    } catch (error) {
      return null;
    }
  }

  private getWsUrl(): string {
    return `${this.wsBaseUrl}${this.port}/`;
  }

  public async start(): Promise<void> {
    if (this.isDestroyed || this.isConnecting) {
      return;
    }

    try {
      const wsUrl = this.getWsUrl();
      this.isConnecting = true;

      this.ws = new WebSocket(wsUrl);
      let timeoutHandle: NodeJS.Timeout | null = null;

      timeoutHandle = setTimeout(() => {
        if (this.ws && this.ws.readyState !== WebSocket.OPEN) {
          try {
            this.ws.removeAllListeners();
            this.ws.terminate();
          } catch (error) {
            // 忽略错误
          }
          
          this.ws = null;
          this.isConnecting = false;
          
          if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.reconnectAttempts++;
            setTimeout(() => {
              this.start().catch(() => {});
            }, 1000);
          }
        }
      }, this.connectionTimeout_ms);

      this.ws.on('open', () => {
        if (timeoutHandle) {
          clearTimeout(timeoutHandle);
          timeoutHandle = null;
        }
        if (this.isDestroyed || !this.ws) return;

        this.isConnecting = false;
        this.reconnectAttempts = 0;

        const loginMsg = `type@=loginreq/roomid@=${this.roomId}/`;
        this.ws.send(this.encodeMsg(loginMsg));

        const joinMsg = `type@=joingroup/rid@=${this.roomId}/gid@=1/`;
        this.ws.send(this.encodeMsg(joinMsg));

        if (this.heartbeatInterval) {
          clearInterval(this.heartbeatInterval);
        }
        this.heartbeatInterval = setInterval(() => {
          if (this.isDestroyed || !this.ws) return;
          if (this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(this.encodeMsg('type@=mrkl/'));
          }
        }, this.heartbeatInterval_ms);
      });

      this.ws.on('message', (data: Buffer) => {
        if (this.isDestroyed) return;

        const msg = this.decodeMsg(data);
        if (!msg || msg.type !== 'chatmsg') return;

        const danmaku: DanmakuMessage = {
          nickname: msg.nn || 'unknown',
          content: msg.txt || '',
          level: msg.level || '0',
          badgeName: msg.bnn || '',
          badgeLevel: msg.bl || '0'
        };

        this.emit('danmaku', danmaku);
      });

      this.ws.on('close', () => {
        if (timeoutHandle) {
          clearTimeout(timeoutHandle);
          timeoutHandle = null;
        }
        if (!this.isDestroyed && this.reconnectAttempts < this.maxReconnectAttempts) {
          this.reconnectAttempts++;
          setTimeout(() => {
            this.start().catch(() => {});
          }, 1000);
        }
      });

      this.ws.on('error', (error) => {
        // 不在这里处理错误，让 close 事件处理重连
      });

    } catch (error) {
      this.isConnecting = false;
      if (!this.isDestroyed && this.reconnectAttempts < this.maxReconnectAttempts) {
        this.reconnectAttempts++;
        setTimeout(() => {
          this.start().catch(() => {});
        }, 1000);
      }
    }
  }

  public stop(): void {
    this.isDestroyed = true;
    this.isConnecting = false;
    
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }
    
    if (this.ws) {
      this.ws.removeAllListeners();
      
      if (this.ws.readyState === WebSocket.OPEN) {
        try {
          this.ws.close();
        } catch (error) {
          // 忽略错误
        }
      }
      
      this.ws = null;
    }
    
    this.removeAllListeners();
    this.reconnectAttempts = 0;
  }

  disconnect() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    
    this.removeAllListeners();
    this.reconnectAttempts = 0;
  }
}

export default DanmakuClient; 

File: src-tauri/src/danmaku.rs
use std::collections::HashMap;
use tokio_tungstenite::{connect_async_tls_with_config, tungstenite::Message};
use futures_util::{SinkExt, StreamExt};
use tokio::time::Duration;
use url::Url;
use tokio::sync::mpsc;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tauri::{Window, Emitter};

pub struct DanmakuClient {
    room_id: String,
    window: Window,
}

impl DanmakuClient {
    pub fn new(room_id: &str, window: Window) -> Self {
        Self {
            room_id: room_id.to_string(),
            window,
        }
    }

    fn encode_msg(&self, msg: &str) -> Vec<u8> {
        let msg_bytes = msg.as_bytes();
        let packet_len = msg_bytes.len() + 9;
        
        let mut result = Vec::new();
        result.extend_from_slice(&(packet_len as u32).to_le_bytes());
        result.extend_from_slice(&(packet_len as u32).to_le_bytes());
        result.extend_from_slice(&689u16.to_le_bytes());
        result.push(0);
        result.push(0);
        result.extend_from_slice(msg_bytes);
        result.push(0);
        
        result
    }

    pub async fn start(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let url = Url::parse("wss://danmuproxy.douyu.com:8506/")?;
        let mut request = url.into_client_request()?;
        request.headers_mut().insert(
            "Sec-WebSocket-Protocol",
            "binary".parse()?
        );

        let (ws_stream, _) = connect_async_tls_with_config(
            request,
            None,
            false,
            None
        ).await?;

        let (mut write, mut read) = ws_stream.split();

        // 发送登录请求
        let login_msg = format!("type@=loginreq/roomid@={}/", self.room_id);
        let login_data = self.encode_msg(&login_msg);
        write.send(Message::Binary(login_data)).await?;

        // 发送加入房间请求
        let join_msg = format!("type@=joingroup/rid@={}/gid@=1/", self.room_id);
        let join_data = self.encode_msg(&join_msg);
        write.send(Message::Binary(join_data)).await?;

        // 创建消息通道
        let (tx, mut rx) = mpsc::channel(32);
        
        // 启动心跳任务
        let heartbeat_msg = "type@=mrkl/";
        let heartbeat_data = self.encode_msg(heartbeat_msg);
        let tx_clone = tx.clone();
        
        tokio::spawn(async move {
            loop {
                tokio::time::sleep(Duration::from_secs(45)).await;
                if let Err(_) = tx_clone.send(Message::Binary(heartbeat_data.clone())).await {
                    break;
                }
            }
        });

        // 消息发送任务
        let send_task = tokio::spawn(async move {
            while let Some(msg) = rx.recv().await {
                if let Err(_) = write.send(msg).await {
                    break;
                }
            }
        });

        let window = self.window.clone();

        // 处理接收到的消息
        while let Some(msg) = read.next().await {
            match msg {
                Ok(Message::Binary(data)) => {
                    if data.len() < 13 {
                        continue;
                    }

                    let content = String::from_utf8_lossy(&data[12..data.len()-1]);
                    let mut result = HashMap::new();
                    for item in content.split('/') {
                        if item.is_empty() {
                            continue;
                        }
                        if let Some((key, value)) = item.split_once("@=") {
                            result.insert(
                                key.to_string(),
                                value.replace("@S", "/").replace("@A", "@")
                            );
                        }
                    }

                    if result.get("type").map_or(false, |t| t == "chatmsg") {
                        let unknown = "unknown".to_string();
                        let empty = "".to_string();
                        let zero = "0".to_string();
                        
                        let danmaku = serde_json::json!({
                            "nickname": result.get("nn").unwrap_or(&unknown),
                            "content": result.get("txt").unwrap_or(&empty),
                            "level": result.get("level").unwrap_or(&zero),
                            "badgeName": result.get("bnn").unwrap_or(&empty),
                            "badgeLevel": result.get("bl").unwrap_or(&zero)
                        });

                        let _ = window.emit("danmaku", danmaku);
                    }
                }
                Ok(Message::Close(_)) | Err(_) => {
                    break;
                }
                _ => {}
            }
        }

        // 取消发送任务
        send_task.abort();

        Ok(())
    }
} 